File: content-solutions/graph-advanced/topo-can-finish.py
from collections import deque


def solution(n, prereqs):
    graph = [[] for _ in range(n)]
    indeg = [0] * n
    for a, b in prereqs:
        graph[a].append(b)
        indeg[b] += 1
    queue = deque(v for v in range(n) if indeg[v] == 0)
    done = 0
    while queue:
        v = queue.popleft()
        done += 1
        for w in graph[v]:
            indeg[w] -= 1
            if indeg[w] == 0:
                queue.append(w)
    return done

File: content-solutions/graph-advanced/topo-can-finish.js
function solution(n, prereqs) {
  const graph = Array.from({ length: n }, () => []);
  const indeg = new Array(n).fill(0);
  for (const [a, b] of prereqs) {
    graph[a].push(b);
    indeg[b]++;
  }
  const queue = [];
  for (let v = 0; v < n; v++) if (indeg[v] === 0) queue.push(v);
  for (let head = 0; head < queue.length; head++) {
    const v = queue[head];
    for (const w of graph[v]) {
      if (--indeg[w] === 0) queue.push(w);
    }
  }
  return queue.length;
}
